### PR TITLE
Versionlock functionality for dnf5

### PR DESCRIPTION
**Pull request by m-blaha**
Replacement of dnf4 versionlock plugin.
The config file format changed from ambiguous plain text with globs to TOML format.

PR contains also commands for config file manipulation (dnf5 versionlock add/exclude/delete/clear). More sophisticated manipulations (e.g. "lock to major version 5", "lock to epoch 1") are currently not supported from commandline - the user needs to edit the config file manually to achieve this.

- [x] Setup `since` fields to 5.1.13 for new API methods before the merging

Adjustments of current CI tests: https://github.com/rpm-software-management/ci-dnf-stack/pull/1434
Closes: https://github.com/rpm-software-management/dnf5/issues/332.

### CODE REVIEW

File: dnf5/commands/list/list.cpp
Comment (jan-kolarik) [inline dnf5/commands/list/list.cpp:148]: Probably needed also for advisory list, info? And what about plugins like changelog, needs-restarting? Or maybe, if I understand it correctly, we should make sure all existing package queries related to non-transactional operations have the flag present?

File: include/libdnf5/rpm/versionlock_config.hpp
Comment (jan-kolarik) [inline include/libdnf5/rpm/versionlock_config.hpp:60]: Or `const &` ?

File: include/libdnf5/rpm/versionlock_config.hpp
Comment (jan-kolarik) [inline include/libdnf5/rpm/versionlock_config.hpp:100]: Same as before, maybe `const &` would be better? Also for the `get_errors()`.

File: include/libdnf5/rpm/versionlock_config.hpp
Comment (jan-kolarik) [inline include/libdnf5/rpm/versionlock_config.hpp:125]: `const std::vector<VersionlockPackage> &`?

File: dnf5/commands/versionlock/versionlock_add.cpp
Comment (jan-kolarik) [inline dnf5/commands/versionlock/versionlock_add.cpp:100]: Shouldn't we NOT exclude versionlocks for versionlock commands? :)

**Review by jan-kolarik — COMMENTED** [review]
I did an initial review and plan to revisit it, likely tomorrow, to perform testing on the commands.

File: include/libdnf5/rpm/package_sack.hpp
Comment (jan-kolarik) [inline include/libdnf5/rpm/package_sack.hpp:31]: Or just `#include "versionlock_config.hpp"` on L27 as others from the same folder.

**Comment by jan-kolarik** [comment]
I'd consider adding the docs for manual editing of the toml config file or track it as the follow-up issue. [This commit message](https://github.com/rpm-software-management/dnf5/pull/1121/commits/994e95431b6d92060b895869be30c84364440673) is a good start.

**Comment by jan-kolarik** [comment]
I've needed to patch the `rpm.i` bindings file to compile with bindings:
[0001-swig-Add-new-versionlock-bindings.patch](https://github.com/rpm-software-management/dnf5/files/14097450/0001-swig-Add-new-versionlock-bindings.patch)

File: dnf5/commands/versionlock/versionlock_add.cpp
Comment (m-blaha) [reply dnf5/commands/versionlock/versionlock_add.cpp:67]: It's definitely a possibility.
The reason why I introduced the `is_valid` method was to keep also invalid entries when the configuration is written back to the toml file (e.g. after adding a new entry using `dnf5 versionlock add xxx` command).
The invalid entries are most likely a result of manual versionlock.toml editing and I did not want their changes to vanish. Basically, I wanted to keep the door open for users to fix any mistakes they made.

File: dnf5/commands/list/list.cpp
Comment (m-blaha) [reply dnf5/commands/list/list.cpp:148]: Well not all - for example `check-upgrade` does not create a transaction but we still want to keep the results in sync with what `dnf upgrade` would do.
But generally you are right - I'll add the flag to missing places.

File: dnf5/commands/repoquery/repoquery.cpp
Comment (m-blaha) [reply dnf5/commands/repoquery/repoquery.cpp:556]: Actually `--upgrades` filter needs to be handled - with this switch repoquery needs to apply versionlock excludes. Thanks for spotting!

File: include/libdnf5/rpm/package_sack.hpp
Comment (m-blaha) [reply include/libdnf5/rpm/package_sack.hpp:154]: We should. But what should be the value? Given that it's not clear in which version the PR's gonna be released.

File: include/libdnf5/rpm/versionlock_config.hpp
Comment (m-blaha) [reply include/libdnf5/rpm/versionlock_config.hpp:125]: Not that easily. Given that returned reference is used to directly manipulate the vector (add new items, delete some items, even clear the vector completely), it would require more API on VersionlockPackage class.

File: dnf5/commands/versionlock/versionlock_add.cpp
Comment (m-blaha) [reply dnf5/commands/versionlock/versionlock_add.cpp:66]: Indeed it's possible. I tried to rewrite the loops using `std::find_if`. The resulting code was a bit shorter, but I didn't find it more readable (especially due to nested lambdas for find_if predicates). I would prefer the current state.

**Comment by m-blaha** [comment]
@jan-kolarik I've resolved most of you comments. To simplify the review process I added the changes as new commits. But I can squash them after the review if you want me to.

File: include/libdnf5/rpm/package_sack.hpp
Comment (jan-kolarik) [reply include/libdnf5/rpm/package_sack.hpp:154]: I'd change it right before the merge, I'll update the description of the PR if you agree...

File: dnf5/commands/versionlock/versionlock_add.cpp
Comment (jan-kolarik) [reply dnf5/commands/versionlock/versionlock_add.cpp:66]: Sure, I just kinda wanted to think about and consider other options.

File: dnf5/commands/versionlock/versionlock_add.cpp
Comment (jan-kolarik) [reply dnf5/commands/versionlock/versionlock_add.cpp:67]: I see. Let's keep it then.

**Review by jan-kolarik — APPROVED** [review]

**Comment by jan-kolarik** [comment]
LGTM, I believe if there is no other complaint or request, we can prepare the since tags to `5.1.13` and merge ...

**Comment by m-blaha** [comment]
Cool, thanks. I'm gonna add a new commit with `Since 5.1.13` later today.

**Pull request merged**
